src/waveform/renderers/allshader/waveformrendererpreroll.cpp

```diff
@@ -101,45 +127,98 @@ void WaveformRendererPreroll::paintGL() {
-            x += polyPixelWidth;
-        }
+        drawPattern(x,
+                halfBreadth - halfMarkerBreadth,
+                end,
+                halfBreadth + halfMarkerBreadth,
+                (end - x) / markerLength,
+                false);
     }
 
-    DEBUG_ASSERT(m_vertices.size() <= reserved);
+    m_pTexture->release();
 
-    const int vertexLocation = m_shader.attributeLocation("position");
-    const int matrixLocation = m_shader.uniformLocation("matrix");
-    const int colorLocation = m_shader.uniformLocation("color");
+    m_shader.disableAttributeArray(vertexLocation);
+    m_shader.disableAttributeArray(texcoordLocation);
+    m_shader.release();
+}
 
-    m_shader.bind();
-    m_shader.enableAttributeArray(vertexLocation);
+void WaveformRendererPreroll::drawPattern(
+        float x1, float y1, float x2, float y2, float repetitions, bool flip) {
+    // Draw a large rectangle with a repeating pattern of the texture
+    const int numVerticesPerTriangle = 3;
+    const int reserved = 2 * numVerticesPerTriangle;
+    const int repetitionsLocation = m_shader.repetitionsLocation();
+    const int vertexLocation = m_shader.positionLocation();
+    const int texcoordLocation = m_shader.texcoordLocation();
 
-    const QMatrix4x4 matrix = matrixForWidgetGeometry(m_waveformRenderer, false);
+    m_vertices.clear();
+    m_texcoords.clear();
+    m_vertices.reserve(reserved);
+    m_texcoords.reserve(reserved);
+    m_vertices.addRectangle(x1, y1, x2, y2);
+    m_texcoords.addRectangle(flip ? 1.f : 0.f, 0.f, flip ? 0.f : 1.f, 1.f);
+
+    m_shader.setUniformValue(repetitionsLocation, QVector2D(repetitions, 1.0));
 
     m_shader.setAttributeArray(
             vertexLocation, GL_FLOAT, m_vertices.constData(), 2);
-
-    m_shader.setUniformValue(matrixLocation, matrix);
-    m_shader.setUniformValue(colorLocation, m_color);
+    m_shader.setAttributeArray(
+            texcoordLocation, GL_FLOAT, m_texcoords.constData(), 2);
 
     glDrawArrays(GL_TRIANGLES, 0, m_vertices.size());
+}
 
-    m_shader.disableAttributeArray(vertexLocation);
-    m_shader.release();
+void WaveformRendererPreroll::generateTexture(float markerLength, float markerBreadth) {
+    const float devicePixelRatio = m_waveformRenderer->getDevicePixelRatio();
+    m_markerLength = markerLength;
+    m_markerBreadth = markerBreadth;
+
+    const int imagePixelW = static_cast<int>(m_markerLength * devicePixelRatio + 0.5f);
+    const int imagePixelH = static_cast<int>(m_markerBreadth * devicePixelRatio + 0.5f);
+    const float imageW = static_cast<float>(imagePixelW) / devicePixelRatio;
+    const float imageH = static_cast<float>(imagePixelH) / devicePixelRatio;
+
+    QImage image(imagePixelW, imagePixelH, QImage::Format_ARGB32_Premultiplied);
+    image.setDevicePixelRatio(devicePixelRatio);
+
+    const float penWidth = 1.5f;
+    const float offset = penWidth / 2.f;
+
+    image.fill(QColor(0, 0, 0, 0).rgba());
```

Comment:
Does this work? Should avoid a temporary QColor construction.
```suggestion
    image.fill(Qt::transparent);
```

Reply:
that works